src/integrationTest/java/com/mongodb/hibernate/CollectionIntegrationTests.java

```diff
@@ -0,0 +1,387 @@
+                .append("booleans", new BsonArray(List.of(BsonBoolean.TRUE)))
+                .append("strings", new BsonArray(List.of(new BsonString("str"))))
+                .append(
+                        "bigDecimals",
+                        new BsonArray(List.of(new BsonDecimal128(new Decimal128(BigDecimal.valueOf(10.1)))))));
+        loadedItem = sessionFactoryScope.fromTransaction(
+                session -> session.find(ItemWithBasicCollectionValues.class, updatedItem.id));
+        assertEquals(updatedItem, loadedItem);
+    }
+
+    @Override
+    public void injectSessionFactoryScope(SessionFactoryScope sessionFactoryScope) {
+        this.sessionFactoryScope = sessionFactoryScope;
+    }
+
+    private static void assertCollectionContainsExactly(String json) { // VAKOTODO remove?
+        assertThat(mongoCollection.find()).containsExactly(BsonDocument.parse(json));
+    }
+
+    private static void assertCollectionContainsExactly(BsonDocument document) {
+        assertThat(mongoCollection.find()).containsExactly(document);
+    }
+
+    @Entity
+    @Table(name = "items")
+    static class ItemWithBasicArrayValues {
+        @Id
+        int id;
+
+        byte[] bytes;
+        char[] chars;
+        int[] ints;
+        long[] longs;
+        double[] doubles;
+        boolean[] booleans;
+        Character[] boxedChars;
+        Integer[] boxedInts;
+        Long[] boxedLongs;
+        Double[] boxedDoubles;
+        Boolean[] boxedBooleans;
+        String[] strings;
+        BigDecimal[] bigDecimals;
+    }
+
+    @Entity
+    @Table(name = "items")
+    static class ItemWithBasicCollectionValues {
+        @Id
+        int id;
+
+        Collection<Character> chars;
+        Collection<Integer> ints;
+        Collection<Long> longs;
+        Collection<Double> doubles;
+        Collection<Boolean> booleans;
+        Collection<String> strings;
+        Collection<BigDecimal> bigDecimals;
+    }
+
+    @Entity
+    @Table(name = "items")
+    // VAKOTODO support and add a test for it, as well as for arrays; also add tests to the embeddable test classes
+    static class ItemWithStructAggregateEmbeddablesCollectionValue {
+        @Id
+        int id;
+
+        Collection<StructAggregateEmbeddableIntegrationTests.SingleValue> embeddables;
+    }
+
+    @Nested
+    class Unsupported {
+        @Test
+        void testBoxedBytesArrayValue() {
+            var item = new ItemWithBoxedBytesArrayValue();
+            {
+                item.id = 1;
+                item.bytes = new byte[] {1};
+                item.boxedBytes = new Byte[] {2};
+            }
+            // this is Hibernate ORM bug, and it goes away if the `ItemWithBoxedBytesArrayValue.bytes` field is removed
+            assertThatThrownBy(() -> sessionFactoryScope.inTransaction(session -> session.persist(item)))
+                    .isInstanceOf(ClassCastException.class);
+        }
```

Comment:
There is likely a bug in Hibernate ORM causing this behavior.

Reply:
I digged a little bit. The root cause is `BasicTypeRegistry` registries both `byte[]` and `Byte[]` with value of the latter. If we use `legacy` as the `hibernate.type.wrapper_array_handling` property, the issue would be gone.

Not 100% sure but it seems this is as expected. We might need to know every details of the new property introduced since 6.2 to fully grasp.

Reply:
The `org.hibernate.mapping.Property#isValid()` provides some helpful context to justify why they converge `byte[]` and `Byte[]` as below (https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/mapping/Property.java#L293):

<img width="1235" alt="Screenshot 2025-05-22 at 5 11 32 PM" src="https://github.com/user-attachments/assets/17e89864-8116-465c-b16a-26096b738d61" />

So from my understanding, if `hibernate.type.wrapper_array_handling ` property is set to `allow`, `Byte[]` instead of `byte[]` is supposed to be used; otherwise `legacy` property might make more sense.

Still kind of defective but mixture of `byte[]` and `Byte[]` seems inconsistent with the `allow` configuration as well.

Reply:
@NathanQingyangXu, thank you for asking this question: https://hibernate.zulipchat.com/#narrow/channel/132094-hibernate-orm-dev/topic/hibernate.2Etype.2Ewrapper_array_handling.20setting.20issue/with/525325554.

Reply:
Christian created a Hibernate ORM ticket at https://hibernate.atlassian.net/browse/HHH-19573